KEP-1027: API Unions for 1.25 (#3377)
kubernetes/enhancements: keps/sig-api-machinery/1027-api-unions/README.md

### Goals

Provide a simple mechanism for API authors to label fields of a resource as
members of a oneOf, in order to receive standardized:

> * Validation - ensuring only one member field is set (or at most one if

Reviewer: Specifically, I want people to *not have to write such validation*, reducing webhook usage & built-in validation code, ideally.

  desired).

> * Normalization - ensuring the API server can modify the fields of the oneOf to

Reviewer: Is this the goal? Or are we merely trying to detect and fail requests sent by clients who have misunderstood?

Reviewer 2: There's no way for unstructured clients to get the object right without clearing the fields with version skew though.

Reviewer: and the question is, do we force users to update such clients by failing their requests, or do we want to try and repair the requests.

Reviewer 2: Right, in the former case, it means that adding a branch to a union is a backwards incompatible change, which I think is a little unfortunate (how do we even update Kubernetes if we start failing clients? We basically can't, so adding a branch to a union will be forbidden in practice).

Reviewer 3: > Is this the goal? Or are we merely trying to detect and fail requests sent by clients who have misunderstood?

I'm also unsure that this is a goal versus detection and good failure messages.

> There's no way for unstructured clients to get the object right without clearing the fields with version skew though.

An unstructured client has access to all the fields in the union, right?  Is it a question of knowledge?

I think an explanation that include if and how server-side-apply should handle unions may help sway me one way or the other.  I'll keep reading below to see how that is addressed.

Author: I narrowed this description down to just talking about normalizing for clients unaware of fields due to version skew.

  best match the intent of the client, despite the client potentially having
  incomplete information

### Non-Goals

The focus of this KEP is on providing unified validation and normalization of
new union types. Migrating existing unions away from their bespoke validation
logic (e.g validation functions), is an explicit non-goal and will be pursued in

> a separate KEP or later release.

Reviewer: I think we should do at least one of each type of union, as a check that our proposal actually works. I agree completeness is not a goal at this point.

Reviewer: I'd want to see it in alpha, because otherwise I'm not sure we're building the right thing. Building something with a difficult/impossible migration path is also a failure mode.

Reviewer 3: > Building something with a difficult/impossible migration path is also a failure mode.

While migration would be ideal, providing something that work for future types still provides significant value and would allow us to avoid our past mistakes and limitations.

Author: see updated note in graduation criteria

## Proposal

### User Stories (Optional)

<!--
Detail the things that people will be able to do if this KEP is implemented.
Include as much detail as possible so that people can understand the "how" of
the system. The goal here is to make this feel real for users without getting
bogged down.

-->

#### Story 1

> As a CRD owner, I can use simple semantics (such as go tags), to express the

Reviewer: openapi tags?

Reviewer 2: Or `go markers`, go tags means something else.

Kubernetes ecosystem.

How will security be reviewed, and by whom?

How will UX be reviewed, and by whom?

Consider including folks who also work outside the SIG or subproject.
-->

## Design Details

### Discriminator Field

We propose that all new unions maintain a "discriminator". This is a field that
points to which of the other "union member" fields is to be respected as the
truly desired union field in the case that there are any conflicts.

In order to demonstrate the need for the discriminator, we developed an
extensive [test matrix](https://docs.google.com/spreadsheets/d/1dIOOqgrgvMI9b2412eVuRSydEaOxhYOoqk6bUjZOY9c/edit?resourcekey=0-wlOfJTC_EX-qpU680STHMA#gid=3601413) that looks at various configurations of the performing
REST operations on a union where the client or server is unaware of a newly
added field to the union (due to version skew).

We present a [guide doc](https://docs.google.com/document/d/1Wruosjo0ELLl1yxauzpsUjgH2fK9KdgXDmOdJ5sG7Kg/edit?resourcekey=0-8Pwzx6EvsFR7VQoXzCTY4Q) on how to interpret the test matrix, but the major
conclusions are as follows (along with the test case number from the test matrix):

> * (Case #22 and #27) If an unstructured client is unaware of field on the union, but wants to clear

Reviewer: define unstructured first?

  the union entirely (assuming the union is optional), it will have no way of doing
  so without a discriminator. With a discriminator, the client can express its
  intention by setting the discriminator to the empty value and the server can

>   respects it intentions and clear any fields the client is unaware of.

Reviewer: ```suggestion
  respect its intentions and clear any fields the client is unaware of.
```

* (Case #12 and #16) If a structured client is unaware of a field in the union that is set and it
  just wants to echo back the union it received in a get request (such as when
  updating other parts of the object), a client without a discriminator will
  silently drop the currently set field, while a client with the discriminator

>   will not change the discriminator value, indicating to the client that no

Reviewer: ```suggestion
  will not change the discriminator value, indicating to the server that no
```

  changes are desired in the union.
* (Case #34 and #39) If a client sets a union field that the server is not aware of, the server
  will silently drop it and attempt to clear the object of the union field. With
  a discriminator, the server will see the unrecognized discriminator value and
  can fail loudly.
* (Case #23 and #28) When a client goes to set a field it knows of, but a separate field it doesn't
  know about is currently set, the server can simply know to always respect the
  discriminator. Without a discriminator, the server will have to do convoluted
  logic to detect that the previously set field has not been modified and that
  only one of the other union fields has been.

### Go tags

We're proposing a new type of tags for go types (in-tree types, and also
kubebuilder types):

- `// +unionDiscriminator` before a field means that this field is the
  discriminator for the union. This field MUST be a string. This field MUST be
  required.

> - `// +unionAllowEmpty` before a discriminator field means that by setting the

Reviewer: The wording here is confusing, I'd go with optional/required as the metaphor personally.

Reviewer 2: Also, I don't think that's needed, we'll always want validation on the discriminator, just allow your discriminator to be empty?

Reviewer: Yes maybe marking the discriminator as optional ("" is never valid, but you can omit it?) may have the same effect as treating the union as optional?

  discriminator to an empty string, none of the union fields are persisted
  (meaning at most one member of the union must be set). If not present on the
  discriminator, exactly one of the union members must be set in order to be
  valid.
- `// +unionMember=<discriminatorName>` before a field means that this

>   field is a member of a union. The `<discriminatorName>` is optional if there

Reviewer: it's also not optional if you have backwards compatibility concerns, I'd consider requiring it personally

Reviewer: Note: I misunderstood what "discriminatorName" meant here. You need to add a way to state what you set the discriminator to in order to select this field.

  a discriminator, the server will see the unrecognized discriminator value and
  can fail loudly.
* (Case #23 and #28) When a client goes to set a field it knows of, but a separate field it doesn't
  know about is currently set, the server can simply know to always respect the
  discriminator. Without a discriminator, the server will have to do convoluted
  logic to detect that the previously set field has not been modified and that
  only one of the other union fields has been.

### Go Markers

We're proposing a new type of tags for go types (in-tree types, and also
kubebuilder types):

- `// +unionDiscriminator` before a field means that this field is the
  discriminator for the union. This field MUST be a string. This field MUST be
  required.
- `// +unionOptional` before a discriminator field means that by setting the
  discriminator to an empty string, none of the union fields are persisted
  (meaning at most one member of the union must be set). If not present on the
  discriminator, exactly one of the union members must be set in order to be
  valid.
- `// +unionMember=<memberName>,discriminatedBy=<discriminatorName>` before a field means that this

>   field is a member of a union. The `<memberName>` is the name of the field that must be set as the discriminator value. It defaults to the json representation of the field name if not specified. The `<discriminatorName>` is optional if there

Reviewer 3: CamelCase is what we've used for enums in the past.

>   field is a member of a union. The `<memberName>` is the name of the field that must be set as the discriminator value. It defaults to the json representation of the field name if not specified. The `<discriminatorName>` is optional if there

Author: Added bit about defaulting to CamelCase

### Goals

Provide a simple mechanism for API authors to label fields of a resource as
members of a oneOf, in order to receive standardized validation and
normalization, rather than having to author it themeselves per
resource as currently done as a workaround in various validation
functions (e.g. `pkg/apis/<group>/validation/validation.go`).

* Validation - ensuring only one member field is set (or at most one if
  desired).
* Normalization - ensuring the API server can modify the fields of the oneOf to
  best match the intent of the client, despite the client potentially having
  incomplete information

> Another goal is to migrate at least one existing union to the new semantics in

Reviewer 3: I'm  not completely sure of this at the moment.  I might be willing to accept a "from now on" state without requiring a fix to previous mistakes.

Author: Removed `this` and added a bit to the testing and graduation criteria based on Daniel's soft acceptance of testing as sufficient for validating the possibility of migrations.

  It MUST correspond to one of the valid enum values of the discriminator's enum
  type. It defaults to the `CamelCase` representation of the field name if not specified. The `<discriminatorName>` is optional if there
  is only union in a struct and required if there are multiple unions per
  struct. It should be the `CamelCase` representation of the field tagged with
  `unionDiscriminator`.

#### Discriminator Values

Here we present a description of how discriminators and their valid values
should be defined.

As described above, the discriminator field must be a string and required.
Because, there are only a few specific values that the discriminator can be, we
propose that all discriminators should be defined as an enum, and should be
tagged so via the enum go marker `// +enum`.

Required unions will have the number of valid discriminator values equal to the
number of member fields (see exception below on empty union members). Optional
unions will have the number of valid discriminator values equal to the number of
member fields, plus one additional value for when "no member" is desired. By

> convention, this "no member" discriminator value should be the empty string.

Reviewer 4: Is it possible to have multiple "no member" options? For example I had a union previously where there were multiple options for the discriminator that currently required no configuration, but maybe did in the future. For example, what if the [DeploymentStrategy](https://github.com/kubernetes/api/blob/edebc6706ddef7e828b5b606fbd0c91c6843c2fa/apps/v1/types.go#L397) had more values for the strategy type? The StrategyType has a semantic value as a discriminant but maybe only `RollingUpdate` actually requires configuration. Does this count as an optional union?

Is this just covered by the empty union members case below?

### Goals

The goal is to enable a union or "oneof" semantics in Kubernetes types, both for

> in-tree types and for CRDs. Validation of these unions should be centralized

Reviewer 2: It's also doing normalization, not just validation.

REST operations on a union where the client or server is unaware of a newly
added field to the union (due to version skew).

We present a [guide doc](https://docs.google.com/document/d/1Wruosjo0ELLl1yxauzpsUjgH2fK9KdgXDmOdJ5sG7Kg/edit?resourcekey=0-8Pwzx6EvsFR7VQoXzCTY4Q) on how to interpret the test matrix, but the major
conclusions are as follows (along with the test case number from the test matrix):

* (Case #22 and #27) If an unstructured client (i.e. a client that represents data as raw json maps with no knowledge of the schema)
  is unaware of field on the union, but wants to clear
  the union entirely (assuming the union is optional), it will have no way of doing
  so without a discriminator. With a discriminator, the client can express its
  intention by setting the discriminator to the empty value and the server can
  respect its intentions and clear any fields the client is unaware of.
* (Case #12 and #16) If a structured client is unaware of a field in the union that is set and it
  just wants to echo back the union it received in a get request (such as when
  updating other parts of the object), a client without a discriminator will

>   silently drop the currently set field, while a client with the discriminator

Reviewer 5: > Originally I was thinking a structured client would drop the field but leave the discriminator unchanged

yes, this is what I would expect

> Either way, clearing the union should require setting the discriminator to the explicit value that corresponds to "no-option" selected that we should ensure is not the default that the enum sets to if the client doesn't recognize the value.

That's not what I mean. Some union types don't require the member field be set. Deployment strategy is a good example... you can set `type: RollingUpdate` and leave the `rollingUpdate` field unset and get rolling update defaults. That is *not* the same as `type: None`

  discriminator. Without a discriminator, the server will have to do convoluted
  logic to detect that the previously set field has not been modified and that
  only one of the other union fields has been.

### Go Markers

We're proposing a new type of tags for go types (in-tree types, and also
kubebuilder types):

- `// +unionDiscriminator` before a field means that this field is the
  discriminator for the union. This field MUST be an enum defined as a string (see section on
  discriminator values). This field MUST be required if there is no default
  option, omitempty if the default option is the empty string, or optional and
  omitempty if a default value is specified with the `// +default` marker.
- `// +unionMember=<memberName> before a field means that this
  field is a member of a union. The `<memberName>` is the name of the field that will be set as the discriminator value.
  It MUST correspond to one of the valid enum values of the discriminator's enum
  type. It defaults to the go (i.e `CamelCase`) representation of the field name if not specified.
  `<memberName>` should only be set if authors want to customize how the fields

>   are represented in the discriminator field.

Reviewer 5: should we require `memberName` to match the serialized JSON name of the field case-insensitively?

number of member fields (see exception below on empty union members). Optional
unions will have the number of valid discriminator values equal to the number of
member fields, plus one additional value for when "no member" is desired. By
convention, this "no member" discriminator value should be the empty string.

We define optional unions as union where "at most" one member field of the union
must be non-nil (as opposed to a required union, where "exactly" one member
field of the union must be non-nil).

##### Empty Union Members

In some cases there are more discriminator values than there are member fields
defined in the struct when that specific member requires no configuration. An
example is the `DeploymentStrategy` where it has one member field `rollingUpdate`,
but two valid discriminator values `RollingUpdate` and `Recreate`. By using an
enum as the discriminator value we are able to define values beyond the member
fields in order to accommodate this pattern.

#### Examples

Below is an example of how to define a union based on the above design

```
// +enum
type UnionType string

const (
       FieldA UnionType = "FieldA"

>        FieldB = "FieldB"
>        FieldC = "FieldC"
>        FieldD = "FieldD"
>        FieldNone = ""

Reviewer 5: ```suggestion
       FieldB UnionType = "FieldB"
       FieldC UnionType = "FieldC"
       FieldD UnionType = "FieldD"
       FieldNone UnionType = ""
```

(see https://go.dev/play/p/mLAmKNEpIJU for why this is needed)

  omitempty if a default value is specified with the `// +default` marker.
- `// +unionMember=<memberName> before a field means that this
  field is a member of a union. The `<memberName>` is the name of the field that will be set as the discriminator value.
  It MUST correspond to one of the valid enum values of the discriminator's enum
  type. It defaults to the go (i.e `CamelCase`) representation of the field name if not specified.
  `<memberName>` should only be set if authors want to customize how the fields
  are represented in the discriminator field. `<memberName>` should match the
  serialized JSON name of the field case-insensitively.
- `// +unionDiscriminatedBy=<discriminatorName>` before a member field identifies which
  discriminator (and thus which union) the member field belongs to. Optional
  unless there are multiple unions/discriminators in a single struct. If used,
  it must be the go (i.e. `CamelCase`) representation of the field name tagged
  with `unionDiscriminator`.

#### Discriminator Values

Here we present a description of how discriminators and their valid values
should be defined.

> As described above, the discriminator field must be a string and required.

Author: I don't think there's anything stopping you from defaulting the discriminator, it just must default to one of the values specified by the discriminator enum

> As described above, the discriminator field must be a string and required.

Reviewer 5: agreed, depending on where normalization happens relative to defaulting

* (Case #34 and #39) If a client sets a union field that the server is not aware of, the server
  will silently drop it and attempt to clear the object of the union field. With
  a discriminator, the server will see the unrecognized discriminator value and
  can fail loudly.
* (Case #23 and #28) When a client goes to set a field it knows of, but a separate field it doesn't
  know about is currently set, the server can simply know to always respect the
  discriminator. Without a discriminator, the server will have to do convoluted
  logic to detect that the previously set field has not been modified and that
  only one of the other union fields has been.

### Go Markers

We're proposing a new type of tags for go types (in-tree types, and also
kubebuilder types):

- `// +unionDiscriminator` before a field means that this field is the
  discriminator for the union. This field MUST be an enum defined as a string (see section on
  discriminator values). This field MUST be required if there is no default
  option, omitempty if the default option is the empty string, or optional and
  omitempty if a default value is specified with the `// +default` marker.

> - `// +unionMember=<memberName>,<optional> before a field means that this

Reviewer 5: this might make it clearer that the `,optional` bit is ... optional, and that "optional" is a literal:

```suggestion
- `// +unionMember=<memberName>[,optional] before a field means that this
```

  type. It defaults to the go (i.e `CamelCase`) representation of the field name if not specified.
  `<memberName>` should only be set if authors want to customize how the fields
  are represented in the discriminator field. `<memberName>` should match the
  serialized JSON name of the field case-insensitively. The comma separated
  optional value determines whether or not the member field must be set when the
  discriminator selects it. Meaning, when `optional` is present on a field, the
  discriminator can select the field even if the field is not set. If optional
  is not present in the `unionMember` tag, then the object will fail validation
  if the discriminator selects the field but it is nil. A field can be marked as
  optional without specifying memberName via `// +unionMember,optional`.
- `// +unionDiscriminatedBy=<discriminatorName>` before a member field identifies which
  discriminator (and thus which union) the member field belongs to. Optional
  unless there are multiple unions/discriminators in a single struct. If used,
  it must be the go (i.e. `CamelCase`) representation of the field name tagged

>   with `unionDiscriminator`.

Author: Yea, I don't see any evidence that we need this. I've dropped it from the design.

Note that this proposes unions to be "at most one of". Whether exactly one is
supported or not should be implemented by the validation logic.

### User Stories (Optional)

<!--
Detail the things that people will be able to do if this KEP is implemented.
Include as much detail as possible so that people can understand the "how" of
the system. The goal here is to make this feel real for users without getting
bogged down.

TODO: ask antoine if these stories are complete/sufficient
-->

#### Story 1

> As a CRD owner, I should be able to author my CRD such that the openapi schema

Reviewer 2: I'd split into these two stories:
1. CRD Authors don't have to write the validation
2. Type users can commit their intent more easily, especially when a new field they don't know has been introduced.

What are the caveats to the proposal?
What are some important details that didn't come across above?
Go in to as much detail as necessary here.
This might be a good place to talk about core concepts and how they relate.
-->

### Risks and Mitigations
- We need to ensure we do not break existing union types. This can be done by
  not forcing existing unions to conform to the newly proposed union semantics.
  Integration testing with older types should give us the confidence to be sure
  we have done so.
- There is a lot of risk for errors when there exists skew between clients and
  server. In the section on normalization, we discuss mitigating these risks.

<!--
What are the risks of this proposal, and how do we mitigate? Think broadly.
For example, consider both security and how this will impact the larger
Kubernetes ecosystem.

How will security be reviewed, and by whom?

How will UX be reviewed, and by whom?

Consider including folks who also work outside the SIG or subproject.
-->

## Design Details

### Go tags

We're proposing a new type of tags for go types (in-tree types, and also
kubebuilder types):

- `// +unionDiscriminator` before a field means that this field is the

>   discriminator for the union. This field MUST be a string. Can be optional or

Reviewer 2: Optional discriminator is confusing, I'd always make it required, though you can allow empty discriminator to mean unset.

}

// This will generate one union, with two fields and a discriminator.
type Union struct {
	// +unionDiscriminator
	// +optional
	UnionType string `json:"unionType"`

	// +optional
        // +unionMember
	FieldA int `json:"fieldA"`

>         // +optional

Reviewer 2: missing `// +unionMember`

}

// This will generate one union, with two fields and a discriminator.
type Union struct {
	// +unionDiscriminator
        // +unionAllowEmpty
        // +required
	UnionType string `json:"unionType"`

        // +unionMember
	// +optional
	FieldA int `json:"fieldA"`
        // +unionMember
        // +optional
	FieldB int `json:"fieldB"`
}

> // This generates two unions, each with two fields and a discriminator. The 

Reviewer: how is this two unions?

Author: copy/pasta error, fixed.

type Union2 struct {
	// +unionDiscriminator
        // +required
	Type string `json:"type"`

> 	// +unionMember=type

Reviewer: ...I was expecting this to show the value which you must set the discriminator to in order to choose this leg of the union, not to say which discriminator this field belongs with.

Author: clarified above

### Discriminator

// TODO: ask Antoine

> For backward compatibility reasons, discriminators should be added to existing

Reviewer 2: Most of these paragraphs aren't accurate anymore.

will be interpreted as an intention to clear all the other fields. See section
below.

#### "At most one" versus "exactly one"

// TODO: check this
We propose supporting both "at most one" and "exactly one" semantics. 

To distinguish between the two, API authors should tag the discriminator as

> "optional" or "required" based on the desired behavior. We the discriminator is

Reviewer 2: As mentioned above, required while allowing empty string (and no omitempty)

set to an empty string (or missing from the object), this will indicate to the
server to clear all fields of the union (or fail the request if the
discriminator is required).

### Normalization and Validation

Normalization refers to the process by which the API server attempts to understand
and correct clients which may provide the server with conflicting or incomplete
information about a union.

Issues primarily arise here because of version skew between a client and server,
such as when a client is unaware of new fields added to a union and thus doesn't
know how to clear these new fields when trying to set a different.

For unions with a discriminator (all newly created unions and some existing
unions), normalization is simple: the server should always respect the
discriminator.

This means two things:
1. when the server receives a request with a discriminator set to a
given field, it should clear out any other fields that are set.

> 2. when the server receives a request with a discriminator set to a given field,

Reviewer 2: Really not convinced about this, this specific feature requires doing the update while looking at the old object. Also, dropping fields is part of more general problem with structure deserializing, why would we fix this just for unions?

Author: So I think the current consensus is we want to error loudly here. Is this right @deads2k?

Author: Removed this based on sig meeting discussion and decision to error loudly. Testing matrix reflects the desired behavior

   but that given field is empty, it should maintain the value of the field that
   currently exists.

> For union types without a discriminator (only existing unions), normalization is a little more

Reviewer 2: Drop these two paragraphs for now, I think we should focus on discriminated unions for now.

}

// This will generate one union, with two fields and a discriminator.
type Union struct {
	// +unionDiscriminator

>         // +unionAllowEmpty

Reviewer 4: As far as I can tell this tag isn't defined in the go markers, did I miss it?

For unions with a discriminator, normalization is simple: the server should always respect the
discriminator.

This means two things:
1. when the server receives a request with a discriminator set to a
given field, it should clear out any other fields that are set.
2. when the server receives a request with a discriminator set to a given field,
   but that given field is empty, the server should fail with a clear error
   message.

For situations where a typed client is unaware of a new field (that is currently
set) and drops the set field such that no fields are now set, the server will
clear all the fields of the union. This is unavoidable for unions without a
discriminator and a major reason why we recommend all new unions have a
discriminator.

Union types without a discriminator must continue to perform normalization and
validation manually (i.e. per resource in the validation functions), as is
currently done for existing union types.

Objects must be validated AFTER the normalization process.

### Open Questions

A few open questions exist with the design that need to be resolved with SIG API
Machinery before moving forward.

> 1. When a server receives an update request with empty data in the member fields

Reviewer: Please distinguish between empty and null in this paragraph, it changes the answer.

   but a validly set discriminator pointing to the object's union member field
   that was previously set, should the server error and inform the client that
   the field pointed to by the discriminator is currently empty, or should the
   server retain the previous value it knows was set to the field being pointed
   to by the discriminator?

> 2. What should the value of the discriminator be when no field in the union is

Reviewer: IMO it should be unset, and that is only permitted for optional unions.

Reviewer: We should look at currently existing discriminated unions and see if any of them already made a choice? But yes.

Author: I'm finding it difficult to survey all the types to find existing unions. 

Most seem to be non-discriminated, but I'm aware of three discriminated unions `MetricSpec`, `MetricStatus`, and `DeploymentStrategy`. 

None of these are optional unions, but I can keep digging to see if I can find an example of an existing, optional, discriminated union.

   to be set. A couple options inclued an empty string, an field common to all
   union (e.g. "NONE"), or a field specified on a per union basis.
3. How should a server behave when it receives a union with multiple fields set
   and the discriminator pointing to one of them? Reject the request, accept the

>    request but don't clear the fields, or automatically clear the fields not

Reviewer: > accept the request but don't clear the fields

I mean, it can't clear all the fields? did you mean except for the field not chosen by the discriminator?

>    specified by the discriminator.

Reviewer: I think there are multiple cases, and if they're indistinguishable, we have to fail everything.
* client adds field, but doesn't clear distinguisher or other field -> client error, fail request
* client adds field & changes distinguisher, but doesn't clear other field -> maybe client didn't know about other field, clear it for them and succeed

And I'll repeat that I think it's acceptable to look at the old object to make this determination on an update.

I think if you do this on a create we should just fail.

<!--
This question should be filled when targeting a release.
For Alpha, describe what tests will be added to ensure proper quality of the enhancement.

For Beta and GA, add links to added tests together with links to k8s-triage for those tests:
https://storage.googleapis.com/k8s-triage/index.html

- <test>: <link to test coverage>
-->

##### e2e tests

<!--
This question should be filled when targeting a release.
For Alpha, describe what tests will be added to ensure proper quality of the enhancement.

For Beta and GA, add links to added tests together with links to k8s-triage for those tests:
https://storage.googleapis.com/k8s-triage/index.html

We expect no non-infra related flakes in the last month as a GA graduation criteria.

- <test>: <link to test coverage>
-->
 We are considering adding kubectl e2e tests to mimic kubectl users performing
 various operations on objects with union fields.

### Graduation Criteria

#### Alpha Graduation

- CRDs can be created with union fields and are properly validated when
  created/updated.

> - Existing unions that have discriminators, are properly tagged and validated via

Reviewer: I'd say at least one, not all. That can wait for beta.

  cluster required to make on upgrade, in order to maintain previous behavior?
- What changes (in invocations, configurations, API use, etc.) is an existing
  cluster required to make on upgrade, in order to make use of the enhancement?
-->

At first, Only new union types will follow the prescribed guinance, so no upgrades/downgrades are possible for types
that don't exist yet.

### Version Skew Strategy

<!--
If applicable, how will the component handle version skew with other
components? What are the guarantees? Make sure this is in the test plan.

Consider the following in developing a version skew strategy for this
enhancement:
- Does this enhancement involve coordinating behavior in the control plane and
  in the kubelet? How does an n-2 kubelet without this feature available behave
  when this feature is used?
- Will any other components on the node change? For example, changes to CSI,
  CRI or CNI may require updating that component before the kubelet.
-->

See test matrix and commentary about discriminators. It clearly documents how
the server will use the discriminator to understand the client's intention even

> if the client is not aware of all union fields because of version skew.

Reviewer: What about skew with alpha on vs off? I don't think there's anything to say but the text should say this.

Author: added

-->

- [x] Feature gate (also fill in values in `kep.yaml`)
  - Feature gate name: APIUnions
  - Components depending on the feature gate: kube-apiserver
  
Request handlers in the api server will call into union validation and
normalization function from the structured-merge-diff repo when feature is
enabled.

###### Does enabling the feature change any default behavior?

No

<!--
Any change of default behavior may be surprising to users or break existing
automations, so be extremely careful here.
-->

###### Can the feature be disabled once it has been enabled (i.e. can we roll back the enablement)?

<!--
Describe the consequences on existing workloads (e.g., if this is a runtime
feature, can it break the existing applications?).

Feature gates are typically disabled by setting the flag to `false` and
restarting the component. No other changes should be necessary to disable the
feature.

NOTE: Also set `disable-supported` to `true` or `false` in `kep.yaml`.
-->

Yes, requests will simply skip union validation and normalization.

###### What happens if we reenable the feature if it was previously rolled back?

> Requests will resume perform union validation and normalization

Reviewer: the important detail is that there's no persistent state behind this feature.

Author: noted

feature flags will be enabled on some API servers and not others during the
rollout. Similarly, consider large clusters and how enablement/disablement
will rollout across nodes.
-->

###### What specific metrics should inform a rollback?

N/A
<!--
What signals should users be paying attention to when the feature is young
that might indicate a serious problem?
-->

###### Were upgrade and rollback tested? Was the upgrade->downgrade->upgrade path tested?

N/A
<!--
Describe manual testing that was done and the outcomes.
Longer term, we may want to require automated upgrade/rollback tests, but we
are missing a bunch of machinery and tooling and can't do that now.
-->

###### Is the rollout accompanied by any deprecations and/or removals of features, APIs, fields of API types, flags, etc.?

N/A
<!--
Even if applying deprecation policies, they may still surprise some users.
-->

### Monitoring Requirements

<!--
This section must be completed when targeting beta to a release.

For GA, this section is required: approvers should be able to confirm the
previous answers based on experience in the field.
-->

###### How can an operator determine if the feature is in use by workloads?

> Simply creating and updating objects with union fields.

Author: fixed

<!--
Ideally, this should be a metric. Operations against the Kubernetes API (e.g.,
checking if there are objects with field X set) may be a last resort. Avoid
logs or events for this purpose.
-->

###### How can someone using this feature know that it is working for their instance?

> 1. Create a new CRD with a union field

Reviewer: (and no other validation mechanism)

Author: added

2. Apply the CRD
3. Create a CR with an invalid union (multiple fields set, no discriminator

>    set), see if the CR is rejected via union validation

Reviewer: Add: when we write the e2e test, a standard union CRD and test CR will be obtainable from that.

Author: added

  - components listing and/or watching resources they didn't before
  - API calls that may be triggered by changes of some Kubernetes resources
    (e.g. update of object X triggers new updates of object Y)
  - periodic API calls to reconcile state (e.g. periodic fetching state,
    heartbeats, leader election, etc.)
-->

###### Will enabling / using this feature result in introducing new API types?

No

<!--
Describe them, providing:
  - API type
  - Supported number of objects per cluster
  - Supported number of objects per namespace (for namespace-scoped objects)
-->

###### Will enabling / using this feature result in any new calls to the cloud provider?

No
<!--
Describe them, providing:
  - Which API(s):
  - Estimated increase:
-->

###### Will enabling / using this feature result in increasing size or count of the existing API objects?

No
<!--
Describe them, providing:
  - API type(s):
  - Estimated increase in size: (e.g., new annotation of size 32B)
  - Estimated amount of new objects: (e.g., new Object X for every existing Pod)
-->

###### Will enabling / using this feature result in increasing time taken by any operations covered by existing SLIs/SLOs?

> No

Reviewer: In GA (maybe beta) we might expect resource reduction / reliability improvement, since this removes a need for webhooks.

<!--
Look at the [existing SLIs/SLOs].

Think about adding additional work or introducing new steps in between
(e.g. need to do X to start a container), etc. Please describe the details.

[existing SLIs/SLOs]: https://git.k8s.io/community/sig-scalability/slos/slos.md#kubernetes-slisslos
-->

###### Will enabling / using this feature result in non-negligible increase of resource usage (CPU, RAM, disk, IO, ...) in any components?

> No

Reviewer: because we already go through SSA code?

Author: noted.

Machinery before moving forward.

1. What do we do when the discriminator is set (to the previously set value),
   but all union member fields are null? This likely implies that the client cleared
   the union member unintentionally, given that the discriminator value remains.
   Should the server:
   a. Error loudly and inform the client that the field pointed to by the
   discriminator is null?
   b. Retain the previously set member field, present on the old object, absent
   from the new object, but pointed to by the discriminator?
2. What should the value of the discriminator be when no field in the union is
   to be set (i.e for optional unions).
   a. mandate that the discriminator is the empty string for all optional unions?
   b. make the field specified on a per union basis (defined in the go markers)?
   c. mandate that the discriminator be unset (thus make the discriminator an
   optional field for optional unios)?
3. How should a server behave when it receives a union with multiple fields set
   and the discriminator modified to point to the newly set member field?
   a. reject the request,

>    b. accept the request but don't clear any fields

Reviewer: This isn't possible, the object is invalid?

Author: I think I agree, but I'm curious to hear from @deads2k who I think prefers this approach (but I'm not sure how it would work given the point you make here).

Author: Removed this based on sig meeting discussion and decision to clear the fields not chosen by the discriminator. Testing matrix reflects the desired behavior

   c. accept the request and clear all the fields except the one chosen by the
   discriminator
4. Given that we want to migrate at least one existing union to use the new
   semantics, which union should we do? Must we upgrade both a discriminated and
   non-discriminated union for alpha or is only upgrading an existing

>    discriminated union sufficient?

Reviewer: one of each, to discover flaws in the design.

Author: That makes sense, but I think there's argument to be made that we shouldn't even try migrating unions that don't use a discriminator and just accept that they will always use the existing manual validations? thoughts.

I've also added a discussion on some proposed initial types to migrate. It's a little shallow, so I'm open to any improvements on a good methodology for choosing the first types to migrate.

5. What should we default the discriminator value of a member field to if not
   specified in the go markers?
   a. the json representation of the field name
   b. the go representation of the field name
   Should we even give users the ability to define it or just stick to whatever

>    we decide the default is?

Reviewer: We need to look at existing discriminated unions. Whatever they do needs to continue to be possible. And ideally they are consistent so we can make it the default too. If everything is already consistent we can mandate the behavior, but that risks leaving CRDs which we don't know about and then can't work.

Author: Of the three discriminated unions I'm aware of (`DeploymentStrategy`, `MetricSpec`, and `MetricStatus`) all three define the valid discriminator values as enums (so technicially they are defined on a per-resource basis), but all follow the convention of using the go representation. 

As long as we give the option of user defined member discriminator value, we can assure that we aren't making anything currently done impossible. 

I guess looking at the existing discriminated unions the better question might be "should we even give users a default instead of just always forcing them to define all the values". 

I think having a default is much nicer ergonomics though and stand by the current approach as it is written (default to go representation, give users the option to override)

Author: See updates to "Go Markers" section

The KEP must have a approver from the
[`prod-readiness-approvers`](http://git.k8s.io/enhancements/OWNERS_ALIASES)
team. Please reach out on the
[#prod-readiness](https://kubernetes.slack.com/archives/CPNHUMN74) channel if
you need any help or guidance.
-->

### Feature Enablement and Rollback

<!--
This section must be completed when targeting alpha to a release.
-->

###### How can this feature be enabled / disabled in a live cluster?

<!--
Pick one of these and delete the rest.

Documentation is available on [feature gate lifecycle] and expectations, as
well as the [existing list] of feature gates.

[feature gate lifecycle]: https://git.k8s.io/community/contributors/devel/sig-architecture/feature-gates.md
[existing list]: https://kubernetes.io/docs/reference/command-line-tools-reference/feature-gates/
-->

- [x] Feature gate (also fill in values in `kep.yaml`)
  - Feature gate name: APIUnions
  - Components depending on the feature gate: kube-apiserver

Request handlers in the api server will call into union validation and
normalization function from the structured-merge-diff repo when feature is
enabled.

###### Does enabling the feature change any default behavior?

> No, since all built in unions are currently validated in other ways.

Reviewer 3: Enabling this would mean that CRs would suddenly get more validation.  I suspect there are cases where already invalid data is present in etcd, how will updates to these CRs proceed?

Author: The enums validation logic will be inert for CRs without enum fields. Are you suggesting there will be existing CRs with union fields and those will have bad data? I get the concern below that disabling the feature once enabled and then re-enabling could have allowed bad data in, but I'm not sure I understand this concern about initial enablement breaking existing CRs

Reviewer 3: > Are you suggesting there will be existing CRs with union fields and those will have bad data?

I think it actually becomes pretty likely as this progresses in its lifecycle. People will start using  it, but not every cluster will have the feature enabled.  I'm ok with this happening, but I think we should describe it accurately.

Author: gotcha, that makes sense. Updated to describe this scenario

<!--
Any change of default behavior may be surprising to users or break existing
automations, so be extremely careful here.
-->

###### Can the feature be disabled once it has been enabled (i.e. can we roll back the enablement)?

<!--
Describe the consequences on existing workloads (e.g., if this is a runtime
feature, can it break the existing applications?).

Feature gates are typically disabled by setting the flag to `false` and
restarting the component. No other changes should be necessary to disable the
feature.

NOTE: Also set `disable-supported` to `true` or `false` in `kep.yaml`.
-->

Yes, requests will simply skip union validation and normalization.

###### What happens if we reenable the feature if it was previously rolled back?

> Requests will resume performing union validation and normalization; there is no

Reviewer 3: This claim doesn't seem quite right.  When used to enforce validation on CRs, there is persistent state in etcd for CR instances.  Some may have this validation enforced and some may not.

Author: Yes, I guess that would cause your CRs to be invalid once you turn union validation back on, but it seems strange to expect to continue to support those? 

You mentioned briefly that ratcheting is not supported in CRs, do you have any ideas about supporting this, otherwise I can only think that we would have to say your CRs become invalid and you have to fix them to continue modifying them, right?

persistent state behind this feature.

> ###### Are there any tests for feature enablement/disablement?

Reviewer 3: For this KEP, I think integration tests demonstrating what happens when this feature is enabled on CRs that already have persisted data that will no longer be valid and how it impacts the update and updatestatus flows.

Author: Added

with and without the feature, are necessary. At the very least, think about
conversion tests if API types are being modified.

Additionally, for features that are introducing a new API field, unit tests that
are exercising the `switch` of feature gate itself (what happens if I disable a
feature gate after having objects written with the new field) are also critical.
You can take a look at one potential example of such test in:
https://github.com/kubernetes/kubernetes/pull/97058/files#diff-7826f7adbc1996a05ab52e3f5f02429e94b68ce6bce0dc534d1be636154fded3R246-R282
-->

### Rollout, Upgrade and Rollback Planning

<!--
This section must be completed when targeting beta to a release.
-->

###### How can a rollout or rollback fail? Can it impact already running workloads?

N/A

<!--
Try to be as paranoid as possible - e.g., what if some components will restart
mid-rollout?

Be sure to consider highly-available clusters, where, for example,
feature flags will be enabled on some API servers and not others during the
rollout. Similarly, consider large clusters and how enablement/disablement
will rollout across nodes.
-->

###### What specific metrics should inform a rollback?

> N/A

Reviewer 3: if I turned this on in alpha, I think I'd be watching the validation failure rates for create and update.  Can you add the specific metrics for those who are slightly less familiar?

Author: added

type Union struct {
  // +unionDiscriminator
  // +required
  Union1 Union1Type `json:"union1"`

  // +unionMember
  // +unionDiscriminatedBy=Union1
  // +optional
  FieldA int `json:"fieldA"`
  // +unionMember,optional
  // +unionDiscriminatedBy=Union1
  // +optional
  FieldB int `json:"fieldB"`

  // +unionDiscriminator
  // +required
  Union2 Union2Type `json:"union2"`

  // +unionMember=ALPHA,
  // +unionDiscriminatedBy=Union2
  // +optional
  Alpha int `json:"alpha"`
  // +unionMember=BETA,optional
  // +unionDiscriminatedBy=Union2
  // +optional
  Beta int `json:"beta"`
}
```

The OpenAPI x-kubernetes-unions extension will then be attached to the discriminator's property and
deserialized into go structs as follows:
```
// XKubernetesUnions is the top level extension
type XKubernetesUnions struct {
  // FieldMembers are the mapping of all valid discriminator values in a
  // union to the corresponding member field.
  // Discriminator value is the value to which the discriminator is set to
  // in order to indicate that a given member field is the currently set member
  // of the union.
  // MemberField may be nil in the case of empty union members where a valid

>   // discriminator value has no corresponding member field.

Reviewer 5: I'm not sure null map values are expressible with all API requests (I think they get interpreted as "drop this key" in some patch types), so this may need to be an empty value (like `{}`) rather than null, but I'm ok figuring that out during implementation

  FieldA Union1Type = "FieldA"
  FieldB Union1Type = "FieldB"
  FieldC Union1Type = "FieldC"
  FieldD Union1Type = "FieldD"
  FieldNone Union1Type = ""
)
const (
  Alpha Union2Type = "ALPHA"
  Beta Union2Type = "BETA"
  Gamma Union2Type = "GAMMA"
  Null Union2Type = "NULL"
)

// This will generate one union, with two fields and a discriminator.
type Union struct {
  // +unionDiscriminator
  // +required
  Union1 Union1Type `json:"union1"`

  // +unionMember
  // +unionDiscriminatedBy=Union1
  // +optional
  FieldA int `json:"fieldA"`
  // +unionMember,optional
  // +unionDiscriminatedBy=Union1
  // +optional
  FieldB int `json:"fieldB"`

  // +unionDiscriminator
  // +required
  Union2 Union2Type `json:"union2"`

  // +unionMember=ALPHA,
  // +unionDiscriminatedBy=Union2
  // +optional
  Alpha int `json:"alpha"`
  // +unionMember=BETA,optional
  // +unionDiscriminatedBy=Union2
  // +optional
  Beta int `json:"beta"`
}
```

turns into:
```
OpenAPIDefinition{
  Schema: spec.Schema{
    SchemaProps: spec.SchemaProps{
      ... // schema props omitted
    },
    VendorExtensible: spec.VendorExtensible{
      Extensions: spec.Extensions{
        "x-kubernetes-unions": []interface{}{

>           map[string]interface{}{

Reviewer 5: would this be more natural to place in the discriminator property, where we already have the enum?

```yaml
properties:
- name: union1
  enum: ["FieldA","FieldB","None"]
  x-kubernetes-union-members:
    FieldA:
      property: fieldA
      optional: true
    FieldB:
      property: fieldB
    None:
      property: ""
```

Reviewer 5: that would keep us from having to repeat `discriminator` a lot, would let us drop a mostly unused level of structure (since multiple unions in a single struct are very unlikely), and would localize validation of the union member keys and the enum

validation could require an entry in x-kubernetes-union-members for every item in the enum, and could validate the existence of all referenced peer properties

>             "discriminator": "Union1",

Reviewer 5: it doesn't make sense to put the go field name in the schema... all the field names referenced in the openapi schema should be the serialized field names

  - Components depending on the feature gate: kube-apiserver

Request handlers in the api server will call into union validation and
normalization function from the structured-merge-diff repo when feature is
enabled.

###### Does enabling the feature change any default behavior?

No, since all built in unions are currently validated in other ways.

<!--
Any change of default behavior may be surprising to users or break existing
automations, so be extremely careful here.
-->

###### Can the feature be disabled once it has been enabled (i.e. can we roll back the enablement)?

<!--
Describe the consequences on existing workloads (e.g., if this is a runtime
feature, can it break the existing applications?).

Feature gates are typically disabled by setting the flag to `false` and
restarting the component. No other changes should be necessary to disable the
feature.

NOTE: Also set `disable-supported` to `true` or `false` in `kep.yaml`.
-->

Yes, requests will simply skip union validation and normalization.

###### What happens if we reenable the feature if it was previously rolled back?

Custom resources that were skipping union validation when when the feature was

> rolled back may have allowed invalid data to persist. These must be updated to

Reviewer 3: I'm possibly ok with this, but I'm interested in whether @liggitt wants to hardcode a version of ratcheting validation for unions to ease this transition.  I think such a thing could help schema authors.

Reviewer 5: I think union validation should be included in the list of types of schema changes we want to handle ratcheting validation for (https://github.com/kubernetes/kubernetes/issues/94060), but I wouldn't hard-code a special version of ratcheting for this feature.

Reviewer 5: I'm actually more interested in the "I update a CRD to add union validation and have to deal with existing data" user story than the "I disable and then re-enable the feature" story. Does this design cover what is required to do *that* successfully?

Reviewer: We don't have to tag existing unions with non-standard validation. So there's no reason to need a ratchet? We can't change the set of objects that are accepted or rejected, but I think it should be OK for the error messages to change wording.

Reviewer: Don't CRDs have that problem in spades? There are loads of incompatible changes you can make that we don't stop you from making. I don't think we should have to solve that problem for unions.

Author: I've added a section on ratcheting validation to try to spell out that we are not commiting to ratcheting validation (and not implementing it for alpha), but should make it reasonably possible to do in the future if necessary. 

Let me know if this is an okay compromise or can be articulated better/differently

Enabling the feature could cause existing CRs to fail validation if the
correspond CRD has union fields and the existing CRs have invalid unions that
were unvalidated when initially created in a cluster that had the unions feature
disabled.

These CRs will need to be corrected in order to pass validation (or the feature
disabled).

<!--
Any change of default behavior may be surprising to users or break existing
automations, so be extremely careful here.
-->

###### Can the feature be disabled once it has been enabled (i.e. can we roll back the enablement)?

<!--
Describe the consequences on existing workloads (e.g., if this is a runtime
feature, can it break the existing applications?).

Feature gates are typically disabled by setting the flag to `false` and
restarting the component. No other changes should be necessary to disable the
feature.

NOTE: Also set `disable-supported` to `true` or `false` in `kep.yaml`.
-->

Yes, requests will simply skip union validation and normalization.

###### What happens if we reenable the feature if it was previously rolled back?

Custom resources that were skipping union validation when when the feature was

> rolled back may have allowed invalid data to persist. These must be updated to

Reviewer 6: > these must me updated to have valid data

Or what will happen? you get errors retrieving up updating them? api server won't start? we need to know the impact of that and make sure it's manageable.

Author: Updated this. In short, updates should error, retrievals should still succeed

Conversion between OpenAPI v2 and OpenAPI v3 will preserve these fields.

### Normalization and Validation

#### Normalization

Normalization refers to the process by which the API server attempts to understand
and correct clients which may provide the server with conflicting or incomplete
information about a union.

Issues primarily arise here because of version skew between a client and a server,
such as when a client is unaware of new fields added to a union and thus doesn't
know how to clear these new fields when trying to set a different field.

For unions with a discriminator, normalization is simple: the server should always respect the
discriminator.

This means two things:

> 1. when the server receives a request with a discriminator set to a

Reviewer 5: ```suggestion
1. when the server receives an update request with a discriminator set to a
```

given field, and multiple member fields are set it should:
  a. If the discriminator has been modified from the one set in the old object,
  the server should clear all fields except the one pointed to by the discriminator.
  b. If the discriminator is unchanged but a new member field has been set (in
  addition to the existing one that has not been cleared), it should error
  loudly that the client must change the discriminator if it changes any union

>   member fields.

Reviewer 5: while I agree with this result, this bit would go in the validation step, not the normalization step, right?

2. when the server receives a request with a discriminator set to a given field,
   but that given field is empty, the server should fail with a clear error

>    message.

Reviewer 5: isn't this the "empty union members case"? also, this should be in the validation section, not normalization, right?

For situations where a typed client is unaware of a new field (that is currently
set) and drops the set field such that no fields are now set, the server will

> clear all the fields of the union. This is unavoidable for unions without a

Reviewer 5: An example might help here... if no fields are set, what fields remain for the server to clear?

Also, is this clearing being proposed to be done generically? if so, that seems to conflict with the "Union types without a discriminator must continue to perform normalization and validation manually" statement

discriminator and a major reason why we recommend all new unions have a
discriminator.

Union types without a discriminator must continue to perform normalization and
validation manually (i.e. per resource in the validation functions), as is
currently done for existing union types.

#### Validation

Objects must be validated AFTER the normalization process.

For custom resources, union validation will be done at the same point as the
existing structural schema validation that occurs in the custom resource handler.
This ensures that any generic validation changes made to all custom resources (such
as the ratcheting validation discussed below), behaves appropriately with union

> validation.

Reviewer 5: This sounds reasonable to me. We also need to clarify when normalization and validation are done for CRD and built-in types in this pipeline:

1. Decoding
2. Defaulting
3. Conversion
4. Mutating admission
5. PrepareForCreate/PrepareForUpdate
6. Validation
7. Validating admission

I would expect normalization around step 5, and validation around step 6. Does that sound correct? One implication of that is that for built-in APIs, those operations will be done on internal type definitions, not versioned ones.

Author: ok, I think I've spelled this out in the doc, but let me know if it can be worded better.

Reviewer 5: sounds good, thanks

#### Ratcheting Validation

When updating CRDs to support union validation, it is possible that existing CRs
become invalid.

The naive solution is to require existing CRs to be updated to a valid state
before they can be updated again.

This creates many potential landmines, and so ratcheting validation is proposed
as an alternative. Ratcheting validation means that objects will ignore stricter
validation rules if and only if the existing object also fails the stricter
validation for the same reason.

Ratcheting validation for custom resources is a separate effort proposed

> outside of this unions effort. For the initial alpha graduation of unions, we

Reviewer 5: link to https://github.com/kubernetes/kubernetes/issues/94060 as a breadcrumb

Conversion between OpenAPI v2 and OpenAPI v3 will preserve these fields.

### Normalization and Validation

#### Normalization

Normalization refers to the process by which the API server attempts to understand
and correct clients which may provide the server with conflicting or incomplete
information about a union in update or patch requests.

Issues primarily arise here because of version skew between a client and a server,
such as when a client is unaware of new fields added to a union and thus doesn't
know how to clear these new fields when trying to set a different field.

For unions that follow this design, normalization is simple: the server should always respect the
discriminator.

This means that when the server receives an update request with a discriminator set to a
given field, and multiple member fields are set it should clear all fields
except the one pointed to by the discriminator _if and only if_ the
discriminator has been modified. Having multiple fields set, and a discriminator
not modified is invalid and caught later by the validation step (see below).

> For both custom resources and built-in types, we expect union normalization to be

Reviewer 2: That's odd, admission mutation will see a somewhat invalid object then?

Author: Yea, I think one has to come first. I don't think I have a strong opinion if you think the mutating "normalization" should come before admission mutation (whichever comes second will see a somewhat invalid object, right)

Author: Discussed offline, we will keep this wording as is

called by the request handlers shortly after mutating admission occurs.

#### Validation

Objects must be validated AFTER the normalization process.

Some validation situations specific to unions are:
1. When multiple union fields are set and the discriminator is not set we should
   error loudly that the client must change the discriminator if it changes any
   union member fields.

> 2. When the server receiveds a request with a discriminator set to a given

Reviewer 2: Yeah, that's the situation we've been discussing, there probably should be a way out of this, this is somewhat opinionated.

Author: updated slightly, let me know how to improve it
